File: go/islands.go
func dfs(grid [][]byte, row int, col int) {
    if (row < 0 || col < 0 || row >= len(grid) || col >= len(grid[0]) || grid[row][col] == '0') {
        return 
    }
    grid[row][col] = '0'
    dfs(grid, row + 1, col)
    dfs(grid, row - 1, col)
    dfs(grid, row, col + 1)
    dfs(grid, row, col - 1)
    
}

func numIslands(grid [][]byte) int {
    numIslands := 0
    for row := 0; row < len(grid); row++ {
        for col := 0; col < len(grid[0]); col++ {
            if grid[row][col] == '1' {
                numIslands++
                dfs(grid, row, col)
            }
        }
    }
    return numIslands 
}

File: techDaily/islands.py
class Solution(object):
    def inRange(self, grid, r, c):
        numRow, numCol = len(grid), len(grid[0])
        if r < 0 or c < 0 or r >= numRow or c >= numCol:
            return False
        return True

    def numIslands(self, grid):
        count = 0 
        row, col = len(grid), len(grid[0])
        for r in range(row):
            for c in range(col):
                tile = grid[r][c]
                if tile == 1: 
                    count += 1   
                    self.dfs(grid, r, c)
                # tile is the actual number 
        return count

    def dfs(self, grid, row, col):
        # check out of bound conditions 
        if not self.inRange(grid, row, col) or grid[row][col] != 1:
            return 
        # visit by placing a zeor 
        grid[row][col] = 0
        self.dfs(grid, row + 1, col)
        self.dfs(grid, row - 1, col)
        self.dfs(grid, row, col + 1)
        self.dfs(grid, row, col - 1)
        
grid = [[1, 1, 0, 0, 0],
        [0, 1, 0, 0, 1],
        [1, 0, 0, 1, 1],
        [0, 0, 0, 0, 0]]

print(Solution().numIslands(grid))
# 3